### ISSUE TITLE
How can I only compile changed/new files?

### ISSUE THREAD

**Issue by Goldjan** [browserify/browserify#1851]
This script compiles all `.js` files from the `./src` js folder to es6. These are saved individually! The path remains the same, only `./src/js` becomes `./build/js`. This is important! I do not want to have `bundle.js`.

**When I call this task, every time all `.js` files are compiled. However, only currently changed files should be compiled. But how? How can I add this into my task?**

package.json (part of it):

    "paths": {
      "src": {
        "base": "./src/",
        "css": "./src/css/",
        "json": "./src/json/",
        "js": "./src/js/",
        "scss": "./src/scss/"
      },
      "dist": {
        "base": "./public/",
        "css": "./public/css/",
        "js": "./public/js/",
        "material": "./public/js/@material/"
      },
      "build": {
        "base": "./build/",
        "css": "./build/css/",
        "js": "./build/js/",
        "html": "./build/html/"
      }
    },
    "globs": {
      "babelJs": [
        "./src/js/**/*.js"
      ] 
    }

gulpfile.js:

    gulp.task('js-babel', function() {
      console.log($.chalk.yellow.bold('--> Transpiling Javascript via Browserify & Babelify...'));
      let files = $.glob.sync(''+pkg.globs.babelJs+'');
      // map them to  streams
      let tasks = files.map(function(entry) {
          return $.browserify({ entries: [entry] })
              .transform('babelify')
              .bundle()
              .pipe($.vinylSourceStream(entry))
              .pipe($.flatten()) // Important!!!!
              .pipe(gulp.dest(pkg.paths.build.js))
              .pipe($.size({gzip: true, showFiles: true}));
        });
        // merge
        return $.eventStream.merge.apply(null, tasks);
    });

**Comment by ronakrathod18** [comment]
@Goldjan - did you find an answer? I am looking for a solution too.

**Comment by goto-bus-stop** [comment]
browserify looks like the wrong tool to use from that code snippet. You should not use browserify on every source file. You should only call it on your entry point(s) and generate a single bundle (or multiple bundles if they get too big). If you want to compile files with babel, and not create browser bundles that include all dependencies, use gulp-babel instead of browserify. Then use gulp-newer to only compile changed files.

[watchify](https://github.com/browserify/watchify) can be used during development to automatically and quickly rebuild the bundle when files change.

**Comment by ronakrathod18** [comment]
@goto-bus-stop - Thanks for the clarification! Is there a way to still create bundles since I have multiple template files without using browserify? I do want to be able to create one bundle per template.

**Comment by ronakrathod18** [comment]
What I have is below. It is pretty similar to OP's example. Let me know if this makes sense.

 glob('./src/js/views/*.js', function (err, files) {
        if (err)
            cb(err);

        var tasks = files.map(function (entry) {
            return browserify({ entries: [entry] })
                .transform(babelify)
                .bundle()
                .pipe(source(entry))
                .pipe(buffer())
                .pipe(sourcemaps.init({
                     loadMaps: true
                 }))
                .pipe(rename({
                    extname: '.bundle.js'
                }))
                .pipe(sourcemaps.write("./"))
                .pipe(gulp.dest('./dist/scripts'));
        });
        es.merge(tasks).on('end', cb);
    });

**Comment by goto-bus-stop** [comment]
Only calling it on views makes more sense to me than calling it on every file. However you're still generating many bundles with mostly duplicate code. It'll be faster and smaller by doing _something_ like

```js
glob('./src/js/views/*.js', function (err, entries) {
  browserify({ entries: entries, debug: true })
     .plugin(require('factor-bundle'), {
       outputs: entries.map(function (name) { return name.replace(/\.js$/, '.bundle.js') })
     })
     .bundle()
     .on('end', cb)
})
```

that will build all pages in one go. then you probably need a separate gulp task to extract the source maps from the generated files. it might need some tweaking, I haven't actually used factor-bundle myself.

**Comment by ronakrathod18** [comment]
@goto-bus-stop - Thanks, my whole purpose was to speed up the browserify task since it takes almost 35 seconds to compile when watch runs. I believe I will still need the watchify task with the above, correct?

**Comment by goto-bus-stop** [comment]
yes if it's still slow I'd recommend adding another task that does that, but also uses watchify. then you can start that task once, and it will re-bundle quickly as you change files.

**Comment by ronakrathod18** [comment]
@goto-bus-stop - thanks a lot! I will give this a try :)
